src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,78 +34,236 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
+     * @param bestMatchingMilestones
+     * @param limit
+     * @param milestones
+     */
+    private void addLikelyUnmatchedMilestones(List<PickerMilestone> bestMatchingMilestones, int limit,
+                                              List<PickerMilestone> milestones) {
+        if (limit <= 0) return;
+        List<PickerMilestone> likelyUnmatchedMilestones =
+                getLimitedFirstMilestones(bestMatchingMilestones, limit, milestones);
+        bestMatchingMilestones.addAll(likelyUnmatchedMilestones);
     }
 
     /**
-     * Gets the list of milestones that matches the current query
+     * Adds up to {@code limit} number of milestones to {@code bestMatchingMilestones}
+     * These milestones are taken from {@code milestones} if they match subsets of {@code userInput}, and are
+     * not already in {@code bestMatchingMilestones}
      *
-     * @return
+     * This method will do nothing if {@code limit} is not a positive integer
+     * @param bestMatchingMilestones
+     * @param limit
+     * @param userInput
+     * @param milestones
      */
-    public List<PickerMilestone> getMatchingMilestonesList() {
-        return this.currentMilestonesList.stream()
-                .filter(milestone -> !milestone.isMatching())
+    private void addPreviouslyMatchingMilestones(List<PickerMilestone> bestMatchingMilestones, int limit,
+                                                 String userInput, List<PickerMilestone> milestones) {
+        if (limit <= 0) return;
+        List<PickerMilestone> previouslyMatchingMilestones =
+                getLimitedMatchesFromPreviousInput(userInput, bestMatchingMilestones, limit, milestones);
+        bestMatchingMilestones.addAll(previouslyMatchingMilestones);
+    }
+
+    /**
+     * Adds up to {@code limit} number of milestones to {@code bestMatchingMilestones}
+     * These milestones are taken from {@code milestones} if its isMatching() returns true
+     *
+     * This method will do nothing if {@code limit} is not a positive integer
+     * @param bestMatchingMilestones
+     * @param limit
+     * @param milestones
+     */
+    private void addMatchingMilestones(List<PickerMilestone> bestMatchingMilestones, int limit,
+                                       List<PickerMilestone> milestones) {
+        if (limit <= 0) return;
+        List<PickerMilestone> matchingMilestones = getLimitedMatchingMilestones(limit, milestones);
+        bestMatchingMilestones.addAll(matchingMilestones);
+    }
+
+    /**
+     * Gets first {@code limit} milestones from {@code milestones} which are not already in {@code allMilestones}
+     *
+     * Assumption: {@code limit} is not negative
+     * @param currentMilestones
+     * @param limit
+     * @param milestones
+     */
+    private List<PickerMilestone> getLimitedFirstMilestones(List<PickerMilestone> currentMilestones,
+                                                            int limit, List<PickerMilestone> milestones) {
```

Comment:
I don't feel comfortable calling `First` in this method, because it suggests that the input list is ordered that you get the first few. I can't think of a better method name, though.